File: 66.plus-one.java
/*
 * @lc app=leetcode id=66 lang=java
 *
 * [66] Plus One
 */

// @lc code=start

import java.util.Arrays;

class Solution {
    public int[] plusOne(int[] digits) {
        if (Arrays.stream(digits).allMatch(x -> x == 9)) {
            int[] ret = new int[digits.length + 1];
            ret[0] = 1;
            return ret;
        }
        plusOne(digits, digits.length - 1);
        return digits;
    }
    private void plusOne(int[] digits, int start) {
        if (digits[start] == 9) {
            digits[start] = 0;
            plusOne(digits, start - 1);
        } else {
            digits[start] += 1;
        }
    }
}
// @lc code=end

File: 66.plus-one.cpp
/*
 * @lc app=leetcode id=66 lang=cpp
 *
 * [66] Plus One
 */

// @lc code=start

#include <vector>

using namespace std;

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        bool carry = true;
        for (auto iter = digits.rbegin(); iter != digits.rend(); ++iter) {
            int tmp = ((carry) ? 1 : 0) + *iter;
            carry = false;
            if (tmp == 10) {
                *iter = 0; carry = true;
            } else {
                *iter = tmp;
            }
        }
        if (carry) digits.insert(digits.begin(), 1);
        return digits;
    }
};
// @lc code=end
